### ISSUE TITLE
[Bug] Fatal Exception : startForegroundService()

### ISSUE THREAD

**Issue by autodistries** [zyrouge/symphony#275]
### Description

My device is pretty weird and has issues; sometimes, despite things seemingly going well, the system crashes and my phone reboots.
This usually happens pretty rarely, but it often crashes right after I start playing music with Symphony (open -> press random play from tracks menu, but might do the same with others play modes (or no play at all) so I will update this issue accordingly).
I have an app called LogFox which (among others) logs crashes, and it was able to catch a fatal error from Symphone before my phone rebooted:

```
FATAL EXCEPTION: main
Process: io.github.zyrouge.symphony, PID: 2738
android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{92abb4c u0 io.github.zyrouge.symphony/.services.radio.RadioNotificationService}
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2006)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7664)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:594)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)

```

(Weird thing : music was still playing when the exception was caught and right before reboot)

### Steps to Reproduce

1. Unstable device
2. Open Symphony
3. Click the random play button
4. Wait a few secs
5. Get crash log
6. Phone reboots

### Expected Behavior

No fatal exception ? (And thus maybe no phone reboot?)

### Device

-   Device: Redmi Note 11 (spesn)
-   OS: Android 11
-   App Version : latest (2023.8.105)

### Additional Context

Issue description has important contextual information
If you feel I could tell you more, I will gladly do. My phone is rooted

### Checklist

-   [x] I have read the [Contribution Guidelines](https://github.com/zyrouge/symphony/wiki/Contributions-Guidelines#issues).

### Updates
- It looks like Symphony crashing does not always triggers reboot : it just didn't (and I got the same datal exception + an ANR this time: 
```
ANR in io.github.zyrouge.symphony
PID: 18178
Reason: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{8d1b910 u0 io.github.zyrouge.symphony/.services.radio.RadioNotificationService}
Load: 0.0 / 0.0 / 0.0
----- Output from /proc/pressure/memory -----
some avg10=0.01 avg60=0.06 avg300=0.07 total=13794676
full avg10=0.00 avg60=0.01 avg300=0.00 total=6165725
----- End output from /proc/pressure/memory -----
CPU usage from 58067ms to 0ms ago (2023-09-16 13:21:49.907 to 2023-09-16 13:22:47.975) with 99% awake:
  48% 18178/usap64: 42% user + 6.6% kernel / faults: 113576 minor 4779 major
  37% 877/surfaceflinger: 21% user + 15% kernel / faults: 713 minor 14 major
  24% 17447/system_server: 13% user + 10% kernel / faults: 216701 minor 26 major
  15% 17728/com.android.systemui: 12% user + 2.7% kernel / faults: 16014 minor 17 major
  12% 18182/usap64: 6.4% user + 5.9% kernel / faults: 31879 minor 5414 major
  11% 757/vendor.qti.hardware.display.composer-service: 7.4% user + 4.4% kernel / faults: 3168 minor
  5.7% 17237/mediaserver: 3.1% user + 2.5% kernel / faults: 4773 minor 5 major
  5.4% 26957/app.revanced.android.apps.youtube.music: 3.3% user + 2.1% kernel / faults: 4165 minor 37 major
  5% 24047/im.molly.app: 3.1% user + 1.8% kernel / faults: 2181 minor
  4.8% 17232/android.hardware.audio.service: 3.7% user + 1.1% kernel / faults: 733 minor 2 major
  4.8% 18228/com.android.launcher3: 3.5% user + 1.2% kernel / faults: 26384 minor 8 major
  4.6% 1247/media.swcodec: 3.3% user + 1.3% kernel / faults: 2349 minor 4 major
  4.2% 18176/usap64: 3.5% user + 0.6% kernel / faults: 8366 minor 3 major
  3.6% 18612/fishnoodle.spacescapewallpaper_free: 2.6% user + 0.9% kernel / faults: 844 minor 3 major
  3.5% 18810/cn.trinea.android.developertools: 2.5% user + 0.9% kernel / faults: 5898 minor 1 major
  2.7% 349/crtc_commit:72: 0% user + 2.7% kernel
  2% 17446/netd: 1.9% user + 0.1% kernel / faults: 176 minor
  1.5% 20702/dev.patrickgold.florisboard.beta: 1.3% user + 0.2% kernel / faults: 2721 minor
  1.5% 17235/audioserver: 1.1% user + 0.3% kernel / faults: 540 minor 6 major
  1.5% 28111/kworker/u16:6: 0% user + 1.5% kernel
  1.4% 265/kgsl_worker_thr: 0% user + 1.4% kernel
  1.3% 761/vendor.qti.hardware.perf@2.2-service: 0.3% user + 0.9% kernel / faults: 47 minor 1 major
  1.3% 18496/irq/52-1b8e300.: 0% user + 1.3% kernel
  1.2% 1/init: 0.3% user + 0.8% kernel / faults: 1616 minor
  1.2% 28109/kworker/u16:2-ufs_clkscaling_0: 0% user + 1.2% kernel
  1.1% 18167/usap64: 0.8% user + 0.2% kernel / faults: 8908 minor 152 major
  1% 493/logd: 0.4% user + 0.6% kernel / faults: 128 minor 16 major
  0.9% 16350/kworker/u16:18-events_unbound: 0% user + 0.9% kernel
  0.8% 151/kswapd0:0: 0% user + 0.8% kernel
  0.8% 495/servicemanager: 0.4% user + 0.4% kernel / faults: 7 minor
  0.7% 10/rcu_preempt: 0% user + 0.7% kernel
  0.6% 17896/com.android.phone: 0.3% user + 0.2% kernel / faults: 608 minor 6 major
  0.6% 18942/org.kde.kdeconnect_tp: 0.3% user + 0.2% kernel / faults: 1406 minor 7 major
  0.6% 29098/kworker/u16:10-ufs_clkscaling_0: 0% user + 0.6% kernel
  0.4% 412/irq/217-fts_ts: 0% user + 0.4% kernel
  0.4% 19449/com.android.providers.media.module: 0.2% user + 0.2% kernel / faults: 3447 minor 7 major
  0.4% 29097/kworker/u16:9-memlat_wq: 0% user + 0.4% kernel
  0.4% 100/system: 0% user + 0.4% kernel
  0.4% 1204/media.extractor: 0.2% user + 0.1% kernel / faults: 2867 minor 19 major
  0.3% 350/crtc_event:72: 0% user + 0.3% kernel
  0.3% 13/rcuop/0: 0% user + 0.3% kernel
  0.3% 32/rcuop/2: 0% user + 0.3% kernel
  0.3% 755/vendor.qti.hardware.display.allocator-service: 0% user + 0.2% kernel / faults: 1040 minor
  0.3% 18865/nextapp.sp: 0.2% user + 0.1% kernel / faults: 12126 minor 8 major
  0.3% 28412/logcat: 0.1% user + 0.1% kernel / faults: 13 minor
  0.2% 496/hwservicemanager: 0% user + 0.1% kernel / faults: 211 minor
  0.1% 28104/kworker/u17:1-kgsl-events: 0% user + 0.1% kernel
  0.1% 29665/kworker/u17:8-kgsl-events: 0% user + 0.1% kernel
  0.1% 18180/usap64: 0% user + 0% kernel / faults: 2831 minor 264 major
  0.1% 24/rcuop/1: 0% user + 0.1% kernel
  0.1% 40/rcuop/3: 0% user + 0.1% kernel
  0.1% 1370/msm_irqbalance: 0% user + 0% kernel / faults: 7 minor
  0.1% 28881/kworker/u17:7-kgsl-events: 0% user + 0.1% kernel
  0.1% 13445/kworker/0:3-pm: 0% user + 0.1% kernel
  0% 48/rcuop/4: 0% user + 0% kernel
  0% 64/rcuop/6: 0% user + 0% kernel
  0% 91/kcompactd0: 0% user + 0% kernel
  0% 461/ueventd: 0% user + 0% kernel / faults: 93 minor
  0% 578/lspd: 0% user + 0% kernel / faults: 47 minor
  0% 744/android.hardware.sensors@2.0-service.multihal: 0% user + 0% kernel / faults: 79 minor
  0% 746/android.hardware.wifi@1.0-service: 0% user + 0% kernel / faults: 102 minor
  0% 16329/kworker/0:6-cgroup_pidlist_destroy: 0% user + 0% kernel
  0% 17864/.qtidataservices: 0% user + 0% kernel / faults: 106 minor 1 major
  0% 17880/com.android.networkstack.process: 0% user + 0% kernel / faults: 248 minor
  0% 17936/android.ext.services: 0% user + 0% kernel / faults: 116 minor 3 major
  0% 19878/com.pittvandewitt.viperfx: 0% user + 0% kernel / faults: 249 minor 2 major
  0% 28639/kworker/3:2-events: 0% user + 0% kernel
  0% 11/rcu_sched: 0% user + 0% kernel
  0% 29/ksoftirqd/2: 0% user + 0% kernel
  0% 72/rcuop/7: 0% user + 0% kernel
  0% 494/lmkd: 0% user + 0% kernel / faults: 10 minor
  0% 530/android.system.suspend@1.0-service: 0% user + 0% kernel / faults: 23 minor
  0% 1292/xtra-daemon: 0% user + 0% kernel / faults: 4 minor
  0% 2158/cds_ol_rx_threa: 0% user + 0% kernel
  0% 27363/eu.thedarken.sdm: 0% user + 0% kernel / faults: 115 minor
  0% 28823/kworker/1:2-events: 0% user + 0% kernel
  0% 56/rcuop/5: 0% user + 0% kernel
  0% 98/ion-pool-uncach: 0% user + 0% kernel
  0% 732/android.hardware.health@2.1-service: 0% user + 0% kernel / faults: 34 minor
  0% 1206/media.metrics: 0% user + 0% kernel / faults: 184 minor
  0% 1319/qcrild: 0% user + 0% kernel / faults: 23 minor
  0% 18174/usap64: 0% user + 0% kernel / faults: 420 minor 6 major
  0% 18966/org.exthmui.game: 0% user + 0% kernel / faults: 137 minor 7 major
  0% 24645/sh: 0% user + 0% kernel / faults: 610 minor
  0% 26748/com.whatsapp: 0% user + 0% kernel / faults: 533 minor 23 major
  0% 26831/org.fdroid.fdroid: 0% user + 0% kernel / faults: 66 minor
  0% 28616/kworker/2:2-events: 0% user + 0% kernel
  0% 29969/android.process.acore: 0% user + 0% kernel / faults: 378 minor 212 major
  0% 2/kthreadd: 0% user + 0% kernel
  0% 9/ksoftirqd/0: 0% user + 0% kernel
  0% 14/rcuos/0: 0% user + 0% kernel
  0% 20/migration/1: 0% user + 0% kernel
  0% 21/ksoftirqd/1: 0% user + 0% kernel
  0% 33/rcuos/2: 0% user + 0% kernel
  0% 44/migration/4: 0% user + 0% kernel
  0% 301/kworker/3:1H-kblockd: 0% user + 0% kernel
  0% 384/core_ctl/4: 0% user + 0% kernel
  0% 500/psimon: 0% user + 0% kernel
  0% 587/ipacm: 0% user + 0% kernel / faults: 53 minor
  0% 710/qrtr_rx: 0% user + 0% kernel
  0% 724/android.hidl.allocator@1.0-service: 0% user + 0% kernel / faults: 230 minor 12 major
  0% 735/android.hardware.memtrack@1.0-service: 0% user + 0% kernel / faults: 10 minor
  0% 776/vendor.qti.hardware.vibrator.service: 0% user + 0% kernel / faults: 6 minor
  0% 988/irq/221-swr_mas: 0% user + 0% kernel
  0% 1220/media.codec: 0% user + 0% kernel / faults: 50 minor
  0% 1224/cnd: 0% user + 0% kernel / faults: 45 minor
  0% 1235/netmgrd: 0% user + 0% kernel / faults: 117 minor 52 major
  0% 1283/qcrild: 0% user + 0% kernel / faults: 261 minor
  0% 1348/busybox: 0% user + 0% kernel / faults: 96 minor
  0% 2157/scheduler_threa: 0% user + 0% kernel
  0% 3192/charge_logger: 0% user + 0% kernel / faults: 5 minor
  0% 3642/tcpdump: 0% user + 0% kernel / faults: 8 minor
  0% 12483/kworker/7:0-mm_percpu_wq: 0% user + 0% kernel
  0% 13884/kworker/5:0-mm_percpu_wq: 0% user + 0% kernel
  0% 14836/kworker/6:2-mm_percpu_wq: 0% user + 0% kernel
  0% 17236/cameraserver: 0% user + 0% kernel / faults: 44 minor 2 major
  0% 17239/wificond: 0% user + 0% kernel / faults: 23 minor
  0% 18410/kworker/u17:0-kgsl-events: 0% user + 0% kernel
  0% 18737/com.android.nfc: 0% user + 0% kernel / faults: 143 minor 10 major
  0% 18876/com.saggitt.omega: 0% user + 0% kernel / faults: 82 minor 10 major
  0% 24515/sh: 0% user + 0% kernel / faults: 393 minor 1 major
  0% 26777/com.android.inputmethod.latin: 0% user + 0% kernel / faults: 214 minor 134 major
  0% 27005/com.mgoogle.android.gms: 0% user + 0% kernel / faults: 309 minor 6 major
  0% 27184/com.snapchat.android: 0% user + 0% kernel / faults: 269 minor
  0% 27889/kworker/4:1-events_power_efficient: 0% user + 0% kernel
  0% 28115/kworker/u17:3-kgsl-events: 0% user + 0% kernel
  0% 28219/kworker/u17:4-kgsl-events: 0% user + 0% kernel
  0% 28957/kworker/0:2-events: 0% user + 0% kernel
 +0% 30093/kworker/6:1: 0% user + 0% kernel
 +0% 30202/kworker/3:0-mm_percpu_wq: 0% user + 0% kernel
 +0% 30210/kworker/2:0-events: 0% user + 0% kernel
29% TOTAL: 18% user + 9.8% kernel + 0.1% iowait + 1.1% irq + 0.3% softirq
CPU usage from 2259776ms to 2259776ms ago (1970-01-01 01:00:00.000 to 1970-01-01 01:00:00.000) with 0% awake:
0% TOTAL: 0% user + 0% kernel
Killing 18178:io.github.zyrouge.symphony/u0a270 (adj 900): bg anr
Completed ANR of io.github.zyrouge.symphony in 825ms, latency 2ms
```
Could this be a device-specific crash ? Maybe because something's overloaded and thus Symphony can't manage to start thus foreground service ?

**Comment by zyrouge** [comment]
> Sorry for my late response.

Why does your device reboot? This is really an unexpected behaviour since an app crashing should not cause the operating system to be affected.

You could read [this](https://stackoverflow.com/questions/44425584/context-startforegroundservice-did-not-then-call-service-startforeground) to get a clear idea of the error.

TL;DR: If the notification is not created within 5 seconds after the foreground service is started, Android would just throw an unhandle-able error which would crash the process no matter what. Android does not guarantee the that the service would be created in 5 seconds either.

**Comment by autodistries** [comment]
Hey, sorry for making you wait this long on this issue !
I've reflashed things and my device has not been restarting when Symphony crashes for about two weeks (so the reboot should not be something to be worried about).
Yeah, it seems like the problem fits 100% that question on StackOverflow.
Do you think a migration from startForegroundService to JobScheduler and JobService (as advised in this [answer](https://stackoverflow.com/a/56534432)) would be feasible ?

**Comment by zyrouge** [comment]
> Hey, sorry for making you wait this long on this issue ! I've reflashed things and my device has not been restarting when Symphony crashes for about two weeks (so the reboot should not be something to be worried about). Yeah, it seems like the problem fits 100% that question on StackOverflow. Do you think a migration from startForegroundService to JobScheduler and JobService (as advised in this [answer](https://stackoverflow.com/a/56534432)) would be feasible ?

Probably not. Even Android's official documentation shows usage of this. We should probably not try to fix what's not broken. Closing this since the problem has resolved for you.
